src/main/java/io/jenkins/plugins/extendedsecuritysettings/JenkinsVersionHeaderFilter.java

```diff
@@ -0,0 +1,119 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package io.jenkins.plugins.extendedsecuritysettings;
+
+import hudson.init.Initializer;
+import hudson.util.PluginServletFilter;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.stream.Collectors;
+
+@Restricted(NoExternalUse.class)
+public class JenkinsVersionHeaderFilter implements Filter {
+
+    @Initializer
+    public static void initialize() throws ServletException {
+        PluginServletFilter.addFilter(new JenkinsVersionHeaderFilter());
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        ServletResponse rsp = isEnabled() ? new HeaderRemovingResponseWrapper((HttpServletResponse) response) : response;
+        chain.doFilter(request, rsp);
+    }
+
+    private boolean isEnabled() {
+        final Jenkins jenkins = Jenkins.getInstanceOrNull();
+        return jenkins != null && ExtendedSecuritySettings.get().isDisableXJenkinsHeaderUnlessAuthorized() &&
+                !jenkins.hasPermission(Jenkins.READ);
+    }
+
+    private static class HeaderRemovingResponseWrapper extends HttpServletResponseWrapper {
+
+        private static final String HEADER_NAME = "X-Jenkins";
+
+        private HeaderRemovingResponseWrapper(HttpServletResponse response) {
+            super(response);
+            response.setHeader(HEADER_NAME, null);
```

Comment:
You probably want to check the `response` is not committed first so that this does not blow up?